File: exercise/python/day22/eid/eid.py
from typing import Union

from eid.parsing_error import ParsingError
from eid.serial_number import parse_serial_number
from eid.sex import parse_sex
from eid.year import parse_year


class EID:
    def __init__(self, sex, year, serial_number):
        self.sex = sex
        self.year = year
        self.serial_number = serial_number

    def check_key(self, potential_key: str) -> bool:
        key = int(potential_key)
        return self.key() == key

    def string_without_key(self) -> str:
        return f"{self.sex.value}{self.year}{self.serial_number}"

    def to_long(self) -> int:
        return int(f"{self.sex.value}{self.year}{self.serial_number}")

    def key(self) -> int:
        return 97 - (self.to_long() % 97)

    def __str__(self) -> str:
        return self.string_without_key() + f"{self.key():02}"


def parse_eid(potential_eid: str) -> Union[ParsingError, EID]:
    sex = parse_sex(potential_eid[0])

    if isinstance(sex, ParsingError):
        return sex

    year = parse_year(potential_eid[1:3])
    if isinstance(year, ParsingError):
        return year

    serial_number = parse_serial_number(potential_eid[3:6])
    if isinstance(serial_number, ParsingError):
        return serial_number

    eid = EID(sex, year, serial_number)

    if eid.check_key(potential_eid[6:]):
        return eid
    else:
        return ParsingError("Invalid key")

File: exercise/java/day22/src/main/java/eid/EID.java
package eid;

import io.vavr.control.Either;
import lombok.AllArgsConstructor;
import lombok.EqualsAndHashCode;
import lombok.experimental.ExtensionMethod;

import java.util.Objects;

import static eid.Sex.parseSex;
import static java.lang.String.format;

@AllArgsConstructor
@EqualsAndHashCode
@ExtensionMethod(StringExtensions.class)
public class EID {
    private final Sex sex;
    private final Year year;
    private final SerialNumber serialNumber;

    public static Either<ParsingError, EID> parse(String potentialEID) {
        return parseSex(potentialEID.charAt(0))
                .map(EIDBuilder::new)
                .flatMap(eidBuilder -> parseYear(potentialEID.substring(1, 3), eidBuilder))
                .flatMap(eidBuilder -> parseSerialNumber(potentialEID.substring(3, 6), eidBuilder))
                .map(eidBuilder -> new EID(eidBuilder.getSex(), eidBuilder.getYear(), eidBuilder.getSerialNumber()))
                .flatMap(eid -> checkKey(potentialEID.substring(6), eid));
    }

    private static Either<ParsingError, EIDBuilder> parseYear(String potentialYear, EIDBuilder builder) {
        return Year.parseYear(potentialYear)
                .map(builder::withYear);
    }

    private static Either<ParsingError, EIDBuilder> parseSerialNumber(String potentialSerialNumber, EIDBuilder builder) {
        return SerialNumber.parseSerialNumber(potentialSerialNumber)
                .map(builder::withSerialNumber);
    }

    private static Either<ParsingError, EID> checkKey(String potentialKey, EID eid) {
        return StringExtensions.toInt(potentialKey)
                .filter(parsedKey -> Objects.equals(eid.key(), parsedKey))
                .map(x -> eid)
                .toEither(new ParsingError("invalid key"));
    }

    @Override
    public String toString() {
        return stringWithoutKey() + format("%02d", key());
    }

    private String stringWithoutKey() {
        return sex.toString() + year + serialNumber;
    }

    private Integer key() {
        return stringWithoutKey()
                .toInt()
                .map(x -> (97 - (x % 97)))
                .get();
    }
}